apollo-router/src/plugins/connectors/request_limit.rs

```diff
@@ -0,0 +1,114 @@
+//! Limits on Connectors requests
+
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+use apollo_federation::sources::connect::ConnectId;
+use parking_lot::Mutex;
+
+/// Key to access request limits for a connector
+#[derive(Eq, Hash, PartialEq)]
+pub(crate) enum RequestLimitKey {
+    /// A key to access the request limit for a connector referencing a source directive
+    SourceName(String),
+
+    /// A key to access the request limit for a connector without a corresponding source directive
+    ConnectorLabel(String),
+}
+
+impl Display for RequestLimitKey {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            RequestLimitKey::SourceName(source_name) => {
+                write!(f, "connector source {}", source_name)
+            }
+            RequestLimitKey::ConnectorLabel(connector_label) => {
+                write!(f, "connector {}", connector_label)
+            }
+        }
+    }
+}
+
+impl From<&ConnectId> for RequestLimitKey {
+    fn from(value: &ConnectId) -> Self {
+        value
+            .source_name
+            .as_ref()
+            .map(|source_name| RequestLimitKey::SourceName(source_name.clone()))
+            .unwrap_or(RequestLimitKey::ConnectorLabel(value.label.clone()))
+    }
+}
+
+/// Tracks a request limit for a connector
+pub(crate) struct RequestLimit {
+    max_requests: usize,
+    total_requests: AtomicUsize,
+}
+
+impl RequestLimit {
+    pub(crate) fn new(max_requests: usize) -> Self {
+        Self {
+            max_requests,
+            total_requests: AtomicUsize::new(0),
+        }
+    }
+
+    pub(crate) fn allow(&self) -> bool {
+        self.total_requests.fetch_add(1, Ordering::Relaxed) < self.max_requests
```

Comment:
Genuine question, is `Relaxed` the right concurrency model for this? From my understanding of the [rust nomicon](https://doc.rust-lang.org/nomicon/atomics.html#relaxed), it seems like the value returned might be reordered, thus (maybe) allowing slightly more requests through than we'd expect. 

Is it OK if the amount of requests that we allow through be slightly higher than the specified max?

(I could also be misreading the nomicon. Feel free to ignore if so)

Reply:
Oh, I see that in your log function you warn if it goes over the max. That's probably fine then. I'll leave the comment above for reference, but feel free to mark as resolved.

Reply:
The counter can go over the max (intentionally, so we can track the total number of attempts), but it will never make more requests than the max. That nomicon entry is confusing - what it really means by "using the counter to synchronize any other accesses" are any places where there are additional memory reads or writes that need to be synchronized. Here the result returned by the atomic operation is compared on the current thread to the immutable max value, which doesn't require additional synchronization.